Added RFC 5 in proposed status (#1)
Vonage/server-sdk-specification: rfcs/proposed/RFC-5-server-sdk-jwt-packages.md

# Server SDK JWT Packages

Updates RFC 1 - Server SDK JWT Packages

## Problem
Many developers, both in and outside of the company, use our APIs without using a full SDK. This can be for many reasons, but authentication is a core issue that all developers using our APIs will face. While key/secret authentication is simple to implement, the ACL-based approach with JWTs provides an additional layer of complexity. There are also competing documentation differences between how JWTs should behave, and a common package can help smooth over those issues.

## Duration
16 Aug 2023 - 30 Aug 2023

## Current State
Proposal

## Proposers
Chris Tankersley

## Detail

> JWT authentication is being used on newer APIs being deployed by Vonage as it provides a finer grain of control over authorization between clients. New applications being developed using APIs such as Messages or Conversations are increasingly being developed by outside teams such as Pre-Sales, and there is a large need for generating JWTs easily. Vonage currently offeres two solutions:

Reviewer: ```suggestion
JWT authentication is being used on newer APIs being deployed by Vonage as it provides a finer grain of control over authorization between clients. New applications being developed using APIs such as Messages or Conversations are increasingly being developed by outside teams such as Pre-Sales, and there is a large need for generating JWTs easily. Vonage currently offers two solutions:
```

* `vonage/cli`, which can generate JWTs

> * `vonage-java-jwt`, which can generate JWTs in Java

Reviewer: ```suggestion
* `vonage-jwt-jdk`, which can generate JWTs in JVM-based languages ([link to repo](https://github.com/Vonage/vonage-jwt-jdk))
```

All of our mainline SDKs already support generating JWTs in some capacity, but generally these JWTs are very wide in capability scope or run under the assumption they are used just for the Voice API. Some SDKs expose JWT generation in a fairly clean-room implementation inside the SDK code itself, while others are tightly coupled to their Vonage Client integrations.

JWTs by themselves are an industry standard (https://jwt.io/) and there are libraries for creating raw JWT objects, but there is additional complexity in crafting the JWT to the expectations of Vonage. By offering a standalone JWT package per-language, we can offer customers a quick development win by making it easy to generate Vonage-compatible JWTs with our requirements, and continue to allow developers and customers the ability to not have to require a “kitchen sink” SDK for one small operation.

## Proposed Solution
### Packaging

> Vonage will provide a self contained JWT token generator for each language, that is automatically installed in the SDKs but can also be installed via each language’s package manager. The package will be stored as an open-source project on Github under `vonage/vonage-jwt-[language]` and will be maintained with the same level of maintenance we provide the full server SDKs.

Reviewer 2: This bit we've already done, right? E.g. https://github.com/Vonage/vonage-jwt-ruby

Author: Correct. Everything outside of the new "Signature Verification" is staying the same

|Claim | exp | - | Unix Timestamp (UTC+0), in seconds, when the JWT expires. Must be at least 30 seconds, maximum of 24 hours, from the iat, and generating using the TTL value. | int | Yes |
| Claim | acl->paths | Path / Paths | Path information for access control to API routes | object | No |
| Claim | sub | Subject | “Subject” or user created and associated with the Vonage Application | string | No |

* - Machine Names may be converted to camelCase, snake_case, or PascalCase as appropriate for the implementing language.

### Signature Verification

Vonage supports signed callbacks via a supplied JWT on incoming webhook/callback payloads. Users MUST be able to verify these JWTs to be able to correctly validate that an incoming payload has originated from the Vonage API servers.

For any service that supports signed callbacks, the incoming request to the customer application MUST contain an `Authorization` header with a `Bearer` token. The user MUST extract the token from the `Authorization` header themselves, and MUST extract the raw string token from the header so that it does not contain the `Bearer` header.

> The SDK implementation MUST include a new method to validation the signature string, and this method SHOULD be named `verifySignature`. It MUST take two parameters, the first being a string representation of the token in base64 format, and the second parameter being the customer's string Secret Key. The method MUST return a boolean value.

Reviewer 2: ```suggestion
The SDK implementation MUST include a new method to validate the signature string, and this method SHOULD be named `verifySignature`. It MUST take two parameters, the first being a string representation of the token in base64 format, and the second parameter being the customer's string Secret Key. The method MUST return a boolean value.
```

> The method MUST default to Sha256 as the signature method, as defined by the Vonage Secure Callbacks implementation.

Reviewer 2: More specifically, isn't it SHA-256 HMAC?

Author: Yup, updated
